app/src/main/java/com/tarmeez/game/auth/LoginFragment.kt

```diff
@@ -0,0 +1,129 @@
+        binding?.userPassword?.hint = getString(R.string.enter_password, " ")
+        binding?.newAccount?.setOnClickListener {
+            binding?.root?.let { view ->
+                Navigation.findNavController(view)
+                    .navigate(R.id.LoginFragmentToRegisterFragment)
+            }
+        }
+        binding?.forgotPassword?.setOnClickListener {
+            binding?.root?.let { view ->
+                Navigation.findNavController(view)
+                    .navigate(R.id.LoginFragmentToForgotPasswordFragment) }
+        }
+        binding?.login?.apply {
+            setOnClickListener {
+                hideKeyBoard()
+                if (validation()) {
+                    authViewModel.login (
+                        binding?.userEmail?.text.toString().trim(),
+                        binding?.userPassword?.text.toString().trim()
+                    )
+                }
+            }
+        }
+        progressDialog = object : AlertDialog(requireContext()){
+            override fun show() {
+                super.show()
+                setContentView(R.layout.loading)
+            }
+        }
+    }
+
+    private fun initObserver() {
+        authViewModel.state.observe(viewLifecycleOwner , { state ->
+            when (state) {
+                is AuthViewModel.State.Authenticated -> {
+                    Log.d(TAG, "User Authenticated")
+                    progressDialog.dismiss()
+                    //Navigate to HomeFragment and create session
+
+                }
+                is AuthViewModel.State.ErrorLogin -> {
+                    progressDialog.dismiss()
+                    Log.d(TAG,  state.message)
+                    //Add snackbasr here to notify the user that something went wrong
+                }
+                is AuthViewModel.State.Loading -> {
+                    progressDialog.show()
+                }
+                else -> {}
+            }
+        })
+    }
+
+    private fun validation():Boolean {
+        var result = true
+        if (binding?.userEmail?.text.toString().isBlank()) {
+            binding?.userEmail?.error = getString(R.string.enter_your_email)
+            result = false
+        }
+        if (binding?.userPassword?.text.toString().isBlank()) {
+            binding?.userPassword?.error = getString(R.string.enter_your_password)
+            result = false
+        }
+        if (!Patterns.EMAIL_ADDRESS.matcher(binding?.userEmail?.text.toString()).matches()) {
+            binding?.userEmail?.error = getString(R.string.invalid_email)
+        }
+        if (binding?.userEmail?.text.toString().isBlank() &&
+            binding?.userPassword?.text.toString().isBlank() ) {
+            binding?.userEmail?.error = getString(R.string.enter_your_email)
+            binding?.userPassword?.error = getString(R.string.enter_your_password)
+            result = false
+        }
+        return result
+    }
+
+    private fun hideKeyBoard() {
```

Comment:
Prefer to move **hideKeyBoard(**) function to another class rather than repeated in fragments
Suggestions: 
1- Using Extensions 
2- Creat a Singleton

Reply:
i moved it into MainActivity